scipy/optimize/_zeros_py.py

```diff
@@ -2866,3 +2866,261 @@ def _differentiate_weights(work, n):
+    direction is downhill. The algorithm repeats from here.
+
+    The new endpoint ``w`` is chosen differently depending on whether or not a
+    boundary `min` or `max` has been set in the downhill direction. Without
+    loss of generality, suppose the downhill direction is to the right, so that
+    ``f(xl) > f(xm) > f(xr)``. If there is no boundary to the right, then ``w``
+    is chosen to be ``xr + factor * (xr - xm)`` where ``factor`` is controlled by
+    the user (defaults to 2.0) so that step sizes increase in geometric proportion.
+    If there is a boundary, ``xmax`` in this case, then ``w`` is chosen to be
+    ``xmax - (xmax - xr)/phi``, with steps slowing to a stop at
+    `xmax`. This cautious approach ensures that a minimum near but distinct from
+    the boundary isn't missed while also detecting whether or not the ``xmax`` is
+    a minimizer when ``xmax`` is reached after a finite number of steps.
+    """  #noqa: E501
+    callback = None  # works; I just don't want to test it
+    temp = _bracket_minimum_iv(func, xm, xl, xr, xmin, xmax, factor, args, maxiter)
+    func, xm, xl, xr, xmin, xmax, factor, args, maxiter = temp
+
+    xs = (xl, xm, xr)
+    xs, fs, args, shape, dtype = _scalar_optimization_initialize(func, xs, args)
+
+    xl, xm, xr = xs
+    fl, fm, fr = fs
+    xmin = np.broadcast_to(xmin, shape).astype(dtype, copy=False).ravel()
+    xmax = np.broadcast_to(xmax, shape).astype(dtype, copy=False).ravel()
+    factor = np.broadcast_to(factor, shape).astype(dtype, copy=False).ravel()
+
+    # To simplify the logic, swap a and b if f(a) < f(b). We should always be marching
+    # downhill in the direction from a to b.
+    comp = fl < fr
+    xl[comp], xr[comp] = xr[comp], xl[comp]
+    fl[comp], fr[comp] = fr[comp], fl[comp]
+    # We only need the boundary in the direction we're traveling.
+    limit = np.where(comp, xmin, xmax)
+
+    unlimited = np.isinf(limit)
+    limited = ~unlimited
+    step = np.empty_like(xl)
+
+    step[unlimited] = (xr[unlimited] - xm[unlimited])
+    step[limited] = (limit[limited] - xr[limited])
+
+    status = np.full_like(xl, _EINPROGRESS, dtype=int)
+    nit, nfev = 0, 3
+
+    work = OptimizeResult(xl=xl, xm=xm, xr=xr, fl=fl, fm=fm, fr=fr, step=step,
+                          limit=limit, limited=limited, factor=factor, nit=nit,
+                          nfev=nfev, status=status, args=args)
+
+    res_work_pairs = [('status', 'status'), ('xl', 'xl'), ('xm', 'xm'), ('xr', 'xr'),
+                      ('nit', 'nit'), ('nfev', 'nfev'), ('fl', 'fl'), ('fm', 'fm'),
+                      ('fr', 'fr')]
+
+    def pre_func_eval(work):
+        work.step[~work.limited] *= work.factor[~work.limited]
+        work.step[work.limited] /= work.factor[work.limited]
+        x = np.empty_like(work.xr)
+        x[~work.limited] = work.xr[~work.limited] + work.step[~work.limited]
+        x[work.limited] = work.limit[work.limited] - work.step[work.limited]
```

Comment:
Thanks, this is more similar to the `_bracket_root` logict now. It makes me realize, though, that I think we could redefine factor at the top like
```python3
work.factor[work.limited] = 1/work.factor[work.limited]
```

and then we don't have do fancy indexing every iteration.

```python3
work.step *= work.factor
```

Reply:
I've changed it for `bracket_minimum`. I think a separate PR for the other function would be better. I had to change

```python
    factor = np.broadcast_to(factor, shape).astype(dtype, copy=False).ravel()
```

to 

```python
    factor = np.broadcast_to(factor, shape).copy().astype(dtype, copy=False).ravel()
```

because `np.broadcast_to` produces a read only view. I guess the `copy=False` might not be needed anymore, but I left it in.